Favicon#34

Nidhi0257 opened the pull request:
resolved bug "Don't link directly to google favicons URL #34"

sliechti commented:
I see two issues, one we are parsing a string and trying to make sense of it, it is much safer to use the ResultSet before the method is called.

The bigger issue is that the method opens a connection for each URL and tries to figure if the domain has a favico, it does this within API methods, meaning that the user's call is waiting in the meanwhile. The other problem is this is not cached so it is done multiple times for a domain we may already know it has no favico.

The right location to do this would be in the fetch cron, while fetching for new entries, it can even be done in a smart way, such that we only look for a favico once a day per domain.

Nidhi0257 commented:
Thank you for the feedback.
Yes I was thinking of this only, favicon are used at multiple places while viewing streams, adding collection, also when we are managing the subscription.
and opening connection multiple times for multiple url will  "HttpURLConnection" will slow down the process. 
I think we should maintain actual url and a flag isFaviconExist in DB, once in day a corn job should check if url contain favicon then we can set the flag true.
this can solve all problem related to resultSet, parsing and waiting time.

Let me know if this sounds good or if any other suggestion

sliechti commented:
> I think we should maintain actual url and a flag isFaviconExist in DB, once in day a corn job should check if url contain favicon then we can set the flag true.

sounds good

Nidhi0257 commented:
As per my analysis, 

1. Favicon used mostly under **reader.jsp** and maximum time used method to view favicon icon is fetchFavIcos() which internally hits an api "/api/v1/sources/get" under SourceAPI class.
Afterlooking into the DB the best place to add boolean column "isFaviconExist" is under **feedsources** table.
2. when user view collection under **collection.jsp**, url is fetched from feedentries table. 
In this case we can fetch url from database using join which will not cost more as I believe page hits will be less.

Let me know can e proceed with this?

sliechti commented:
The first step is finding out if a URL has a .favicon or not, (this is in the new method you created)
We need to put that logic in the cron fetcher so it is done in the background and not on each user call to the API. Let's save the result as in feedsources, but let's save it as faviconUrl, empty if it doesn't exits, or the URL if it exists.

For the view, if favico URL is empty we show a 1x1 pixel, if it is not empty, we set the image src to the favicon URL.

Also: It is possible for a page to declare the favicon to be different, so looking for URL/favicon.ico is not 100% correct.
http://stackoverflow.com/questions/9943771/adding-a-favicon-to-a-static-html-page

Nidhi0257 commented:
Committed the code, Please review.
I have created 1 plain file of 1X1 pixel and added to the branch.
I have added boolean variable b_is_favicon_exist to check whether favicon exist or not instead of string, thought boolean variable can be more efficient.
let me know your suggestion.
